platforms/software/maven/src/main/java/org/gradle/api/publish/maven/internal/artifact/AbstractMavenArtifact.java

```diff
@@ -16,55 +16,52 @@
 
 package org.gradle.api.publish.maven.internal.artifact;
 
-import com.google.common.base.Strings;
 import org.gradle.api.internal.tasks.DefaultTaskDependency;
-import org.gradle.api.publish.internal.PublicationArtifactInternal;
 import org.gradle.api.internal.tasks.TaskDependencyFactory;
+import org.gradle.api.model.ObjectFactory;
+import org.gradle.api.provider.Property;
+import org.gradle.api.provider.Provider;
+import org.gradle.api.provider.ProviderFactory;
+import org.gradle.api.publish.internal.PublicationArtifactInternal;
 import org.gradle.api.publish.maven.MavenArtifact;
 import org.gradle.api.tasks.TaskDependency;
 
-import java.io.File;
-
 public abstract class AbstractMavenArtifact implements MavenArtifact, PublicationArtifactInternal {
     private final TaskDependency allBuildDependencies;
     private final DefaultTaskDependency additionalBuildDependencies;
-    private String extension;
-    private String classifier;
-
-    protected AbstractMavenArtifact(TaskDependencyFactory taskDependencyFactory) {
+    private final Property<String> extensionProperty;
+    private final Property<String> classifierProperty;
+
+    protected AbstractMavenArtifact(
+        TaskDependencyFactory taskDependencyFactory,
+        ObjectFactory objectFactory,
+        ProviderFactory providerFactory
+    ) {
         this.additionalBuildDependencies = new DefaultTaskDependency();
         this.allBuildDependencies = taskDependencyFactory.visitingDependencies(context -> {
             context.add(getDefaultBuildDependencies());
             additionalBuildDependencies.visitDependencies(context);
         });
+        this.extensionProperty = objectFactory.property(String.class);
+        this.classifierProperty = objectFactory.property(String.class);
```

Comment:
Why use object factory to instantiate these instead of defining an abstract getter?